NV 487 Add theme customization in notification-center

djabarovgeorge opened the pull request.

scopsy commented on packages/notification-center/src/components/notification-center/components/AppContent.tsx:
```diff
@@ -1,19 +1,17 @@
   const { isLoggedIn } = useAuth();
-  const { colorScheme } = useContext(ThemeContext);
-
+  const { theme } = useNovuThemeProvider();
```
Nice!

scopsy commented on packages/notification-center/src/components/notification-center/components/AppContent.tsx:
```diff
@@ -22,22 +20,26 @@ export function AppContent() {
-      fontColor: colorScheme === 'light' ? colors.B40 : colors.white,
-      secondaryFontColor: colorScheme === 'light' ? colors.B80 : colors.B40,
+      main: organization?.branding?.color || theme.mainColor,
+      fontColor: theme.layoutWrap.colors.fontColor,
+      secondaryFontColor: theme.layoutWrap.colors.secondaryFontColor,
     },
-    fontFamily: organization?.branding?.fontFamily || 'Lato',
+    fontFamily: organization?.branding?.fontFamily || theme.fontFamily,
     layout: {
       direction: (organization?.branding?.direction === 'rtl' ? 'rtl' : 'ltr') as 'ltr' | 'rtl',
     },
```
I would go with code first approach and flip the fallback to UI config.

scopsy commented on packages/notification-center/src/hooks/use-novu-theme-provider.hook.ts:
```diff
@@ -0,0 +1,11 @@
+import { useContext } from 'react';
+import { ThemeContext } from '../store/novu-theme.context';
+
+export function useNovuThemeProvider() {
+  const { colorScheme, theme } = useContext(ThemeContext);
+
+  return {
+    colorScheme,
+    theme,
+  };
+}
```
Good choice refactoring this!

scopsy commented on packages/notification-center/src/store/novu-theme-provider.context.tsx:
```diff
@@ -0,0 +1,35 @@
+
+export function NovuThemeProvider(props: INovuThemeProviderProps) {
+  const novuTheme =
+    props.theme.colorScheme === 'light'
+      ? { ...defaultLightTheme, ...props.theme.light }
+      : { ...defaultDarkTheme, ...props.theme.dark };
```
Imagine the following use case where I want to override this darktheme:
```
<PopoverComponent theme={{ layoutWrap: { colors: { main: 'white' }}}}></...>
```

I would suggest doing a deepMerge here to make sure only partial updates will be used. 

this will effectively delete all the other props there: "fontColor", "secondaryFontColor" or will force the user to provide all of the data there.

djabarovgeorge replied:
Totally agree I missed it on the conversion to objects

scopsy commented on packages/notification-center/src/shared/config/themeDefaultValues.ts:
```diff
@@ -0,0 +1,49 @@
+      secondaryFontColor: colors.B80,
+    },
+  },
+  layout: { background: colors.white, boxShadow: shadows.medium },
```
Was thinking about if we can have similar colors grouped together such as: backgroundColor, secondaryBackground and etc... 

And than do the following fallback:
1) Use specific theme: theme.header.backgroundColor
2) if 1. was not specified use general theme.backgrounColor
3) if both wasn't specified by user, use a default value of backgroundColor

djabarovgeorge commented on packages/notification-center/src/components/notification-center/NotificationCenter.tsx:
```diff
@@ -14,29 +11,26 @@ export interface INotificationCenterProps {
-        </NotificationCenterContext.Provider>
-      </QueryClientProvider>
-    </ThemeContext.Provider>
+    <QueryClientProvider client={queryClient}>
```
Yea 👍

djabarovgeorge commented:
@scopsy Hey I removed the general configuration and added the lodash merge instead of the mapper.
Would you like me to add the notification-center border param? if so I think we should expand the common prop with structured abstracted objects

scopsy commented on packages/notification-center/src/store/novu-theme.context.ts:
```diff
@@ -1,10 +1,79 @@
+  popover?: IThemePopover;
+  notificationListItem?: IThemeNotificationListItem;
+  footer?: IThemeFooter;
+  mainColor?: string;
```
Can we remove this one? Looks like it's only used for the loader animation as a fallback color, we can maybe have a specific category for the loader.

scopsy commented on packages/notification-center/src/store/novu-theme.context.ts:
```diff
@@ -1,10 +1,79 @@
+  logoPrefixFontColor?: string;
+}
+
+export interface IThemeBellGradientDot {
```
```suggestion
export interface IThemeUnseenBadge {
```

scopsy commented on packages/notification-center/src/store/novu-theme.context.ts:
```diff
@@ -1,10 +1,79 @@
+}
+
+export interface IThemeBellGradientDot {
+  color?: {
+    stopColor?: string;
+    stopColorOffset?: string;
+    borderColor?: string;
+  };
```
```suggestion
  color?: string | {
    stopColor?: string;
    stopColorOffset?: string;
    borderColor?: string;
  };
```

In this case, we can do simple color or a gradient object.

djabarovgeorge replied:
Agree

scopsy commented on packages/notification-center/src/store/novu-theme.context.ts:
```diff
@@ -1,10 +1,79 @@
+export interface INovuTheme {
+  layout?: IThemeLayout;
+  header?: IThemeHeader;
+  popover?: IThemePopover;
```
This also can extend from INovuTheme in a INovuPopoverTheme interface aswell.

scopsy reviewed (commented):
Final comments and ideas here ^ Let me know what you think

djabarovgeorge commented on packages/notification-center/src/components/notification-center/components/AppContent.tsx:
```diff
@@ -22,22 +20,21 @@ export function AppContent() {
-      main: organization?.branding?.color || colors.vertical,
-      fontColor: colorScheme === 'light' ? colors.B40 : colors.white,
-      secondaryFontColor: colorScheme === 'light' ? colors.B80 : colors.B40,
+      main: theme.loaderColor || organization?.branding?.color,
```
@scopsy Is this what you meant?

In addition at the current state, I am not sure there is a purpose of the organization configuration because we use here the theme which is or defined by the user via the code or the defaults that there are so there won't be any fallback to UI config, what do you think?

scopsy replied:
We might leave this out then on the UI part later. Let's leave it like this for now. It's great.
